rpcs3/Emu/Cell/Modules/cellAdec.h

```diff
@@ -471,6 +458,175 @@ struct AdecCmdQueue
+
+	CellAdecAuInfo au_info;
+	CellAdecPcmItem pcm_item;
+
+	const u32 reserved1 = 0;
+	const u32 reserved2 = 0;
+
+	// Frames that are ready to be consumed form a linked list. However, this list is not used (AdecOutputQueue is used instead)
+	be_t<s32> next = 0; // Index of the next frame that can be consumed
+	be_t<s32> prev = 0; // Index of the previous frame that can be consumed
+
+	AdecFrame(s32 index, u32 bitstream_info_addr)
+		: this_index(index)
+	{
+		pcm_item.pcmAttr.bsiInfo.set(bitstream_info_addr);
+	}
+};
+
+CHECK_SIZE(AdecFrame, 0x68);
+
+class AdecOutputQueue
+{
+	struct entry
+	{
+		const be_t<s32> this_index; // Unused
+		be_t<s32> state = 0xff; // 0xff = empty, 0x10 = filled
+		vm::bptr<CellAdecPcmItem> pcm_item = vm::null;
+		be_t<s32> pcm_handle = -1;
+	}
+	entries[4];
+
+	be_t<s32> front = 0;
+	be_t<s32> back = 0;
+	be_t<s32> size = 0;
+
+	shared_mutex mutex; // sys_mutex_t
+	be_t<u32> cond{};   // sys_cond_t, unused
+
+public:
+	AdecOutputQueue() : entries{ {0}, {1}, {2}, {3} } {}
+
+	error_code push(vm::ptr<CellAdecPcmItem> pcm_item, s32 pcm_handle)
+	{
+		std::lock_guard lock{mutex};
+
+		if (entries[back].state != 0xff)
+		{
+			return true; // LLE returns the result of the comparison above
+		}
+
+		entries[back].state = 0x10;
+		entries[back].pcm_item = pcm_item;
+		entries[back].pcm_handle = pcm_handle;
+
+		back = (back + 1) & 3;
+		size++;
+
+		return CELL_OK;
+	}
+
+	const entry* pop()
+	{
+		std::lock_guard lock{mutex};
+
+		if (entries[front].state == 0xff)
+		{
+			return nullptr;
+		}
+
+		const entry* const ret = &entries[front];
+
+		entries[front].state = 0xff;
+		entries[front].pcm_handle = -1;
+
+		front = (front + 1) & 3;
+		size--;
+
+		return ret;
+	}
+
+	const entry& peek()
+	{
+		std::lock_guard lock{mutex};
+		return entries[front];
+	}
+};
+
+CHECK_SIZE(AdecOutputQueue, 0x54);
+
+enum class AdecSequenceState : u32
+{
+	dormant = 0x100,
+	ready = 0x200,
+	closed = 0xa00,
+};
+
+struct AdecContext // CellAdecHandle = AdecContext*
+{
+	vm::bptr<AdecContext> _this;
+	const be_t<u32> this_size; // Size of this struct + AdecFrames + bitstream info structs
+
+	const u32 unk = 0; // Unused
+
+	be_t<AdecSequenceState> sequence_state = AdecSequenceState::dormant;
+
+	const CellAdecType type;
+	const CellAdecResource res;
+	const CellAdecCb callback;
+
+	const vm::bptr<void> core_handle;
+	const vm::bcptr<CellAdecCoreOps> core_ops;
+
+	CellCodecTimeStamp previous_pts{ CODEC_TS_INVALID, CODEC_TS_INVALID };
+
+	const be_t<s32> frames_num;
+	const u32 reserved1 = 0;
+	be_t<s32> frames_head = -1;       // Index of the oldest frame that can be consumed
+	be_t<s32> frames_tail = -1;       // Index of the most recent frame that can be consumed
+	const vm::bptr<AdecFrame> frames; // Array of AdecFrames, number of elements is return value of CellAdecCoreOps::getPcmHandleNum
+
+	const be_t<u32> bitstream_info_size;
+
+	u64 mutex_attribute[5]{}; // sys_mutex_attribute_t
+	shared_mutex mutex;       // sys_mutex_t
```

Comment:
Do not use this type on PS3 structures, you can make it an index to some shared_index array or something